File: LeetCode/Python/group_anagram.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        if len(strs)==0:
            return []
        d = defaultdict(list)
        for word in strs:
            d[''.join(sorted(word))].append(word)

        return d.values()

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        if len(strs)==0:
            return []
        d = defaultdict(list)
        for word in strs:
            d[tuple(sorted(word))].append(word)

        return d.values()
        

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        if len(strs)==0:
            return []
        d = defaultdict(list)
        for word in strs:
            tmp = [0]*26
            for ch in word:
                tmp[ord(ch)-ord('a')]+=1
            d[tuple(tmp)].append(word)

        return d.values()
        

File: LeetCode/Java/GroupAnagram.java
import java.util.*;

public class GroupAnagram {

    public List<List<String>> groupAnagrams(String[] strs) {
        // Method 1
        /*
            HashMap<String, List<String>> map = new HashMap<>();
            for (String word : strs){
                char[] tmp = word.toCharArray();
                Arrays.sort(tmp);
                String sortedWord = Arrays.toString(tmp);
                if (!map.containsKey(sortedWord)){
                    map.put(sortedWord, new ArrayList<>());
                }
                map.get(sortedWord).add(word);

            }
            return new ArrayList<>(map.values());
        */

        HashMap<String, List<String>> map = new HashMap<>();
        for (String word : strs){
            int[] tmp = new int[26];
            for (char c:word.toCharArray()){
                tmp[c-'a']++;
            }
            String sortedWord = Arrays.toString(tmp);
            if (!map.containsKey(sortedWord)){
                map.put(sortedWord, new ArrayList<>());
            }
            map.get(sortedWord).add(word);

        }
        return new ArrayList<>(map.values());
        
    }
    
}
